src/garage/tf/algos/ddpg.py

```diff
@@ -248,6 +257,39 @@ def __setstate__(self, state):
         self.__dict__.update(state)
         self.init_opt()
 
+    def train(self, runner):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+        runner.enable_logging = False
+
+        for _ in runner.step_epochs():
+            for cycle in range(self._steps_per_epoch):
+                runner.step_path = runner.obtain_samples(runner.step_itr)
+                for path in runner.step_path:
+                    path['rewards'] *= self._reward_scale
+                last_return = self.train_once(runner.step_itr,
+                                              runner.step_path)
+                if (cycle == 0 and self.replay_buffer.n_transitions_stored >=
+                        self._min_buffer_size):
+                    runner.enable_logging = True
+                    log_performance(runner.step_itr,
+                                    obtain_evaluation_samples(
+                                        self.policy, runner.get_env_copy()),
```

Comment:
Should it pass a copy of the policy? Otherwise, it might alter policy's state if there's any, e.g. hidden state for recurrent policies, but it might not be an issue since this is called after the training, at the end of a training epoch.

Perhaps a safer approach is to make a policy copy inside `obtain_evaluation_samples()` so it doesn't modify the policy unexpectedly.

Reply:
This is a valid question, though it's worth nothing that we don't have RNN support for off-policy algos yet (it's possible but is a special training procedure)